build_utils/bump_changelog.py

```diff
@@ -0,0 +1,115 @@
+#! /usr/bin/env python
+
+"""Bump the current version number in CHANGELOG.rst
+
+Unreleased changes will be moved into a section for the new version number
+"""
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from pathlib import Path
+import re
+
+from toolz.itertoolz import partition
+
+REPOSITORY_ADDRESS = "https://github.com/pace-neutrons/Euphonic"
+
+
+@dataclass
+class Block:
+    """CHANGELOG data corresponding to a single release"""
+    tag: str
+    previous_tag: str
+    content: str
+
+
+def parse_changelog(changelog_file: Path) -> list[Block]:
+    """Read all sections from changelog file
+
+    Each section should begin with two lines in format
+
+      `TAG <repository_address/compare/PREV_TAG...TAG>`_
+      --------------------------------------------------
+
+    except for 'Unreleased' which compares with HEAD
+    """
+
+    with open(changelog_file, "rt", encoding="utf8") as fd:
```

Comment:
```suggestion
    with changelog_file.open(encoding="utf8") as fd:
```

Reply:
What's the benefit here?

Reply:
Stops `ruff` complaining.

Reply:
https://docs.astral.sh/ruff/rules/builtin-open/
https://docs.astral.sh/ruff/rules/redundant-open-modes/

Reply:
I don't get it. PT123 is meant to be "use Pathlib!" but also enforces this entirely equivalent `open` syntax which "can improve readability"?

I get that `Path(path).open()` is more readable than `open(Path(path))`, but don't see that `path.open()` is more readable than `open(path)`.

Reply:
So the one "advantage" I can see is that it enforces the use of `Path` consistently and will raise if it isn't. Whether this is "good" behaviour is certainly debatable.

Reply:
Will commit it "for fun" then, in this case it doesn't matter much either way.